types/reach__router/index.d.ts

```diff
@@ -0,0 +1,142 @@
+// Type definitions for @reach/router 1.0
+// Project: https://github.com/reach/router
+// Definitions by: Kingdaro <https://github.com/kingdaro>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.8
+
+import * as React from "react";
+
+export type WindowLocation = Window["location"];
+
+export interface History {
+    readonly location: string;
+    readonly transitioning: boolean;
+    listen: (listener: HistoryListener) => HistoryUnsubscribe;
+    navigate: NavigateFn;
+}
+
+export type HistoryListener = () => void;
+export type HistoryUnsubscribe = () => void;
+
+export class Router extends React.Component<RouterProps> {}
+
+export interface RouterProps {
+    basepath?: string;
+    primary?: boolean;
+    location?: WindowLocation;
+}
+
+export type RouteComponentProps<TParams = {}> = Partial<TParams> & {
+    path?: string;
+    default?: boolean;
+    location?: WindowLocation;
+    navigate?: NavigateFn;
+    uri?: string;
+};
+
+export type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;
+
+export type AnchorProps = Omit<
+    React.DetailedHTMLProps<
+        React.AnchorHTMLAttributes<HTMLAnchorElement>,
+        HTMLAnchorElement
+    >,
+    "href" // remove href, as it's ignored by the router
+>;
+
+export interface LinkProps<TState> extends AnchorProps {
+    to?: string;
+    replace?: boolean;
+    getProps?: (props: LinkGetProps) => {};
+    state?: TState;
+}
+
+export interface LinkGetProps {
+    isCurrent: boolean;
+    isPartiallyCurrent: boolean;
+    href: string;
+    location: WindowLocation;
+}
+
+export class Link<TState> extends React.Component<LinkProps<TState>> {}
+
+export interface RedirectProps {
+    from?: string;
+    to: string;
+    noThrow?: boolean;
+}
+
+export class Redirect extends React.Component<RedirectProps> {}
+
+export interface MatchProps<TParams> {
+    path: string;
+    children: MatchRenderFn<TParams>;
+}
+
+export type MatchRenderFn<TParams> = (
+    props: MatchRenderProps<TParams>,
+) => React.ReactNode;
+
+export interface MatchRenderProps<TParams> {
+    match: null | { uri: string; path: string } & TParams;
+    location: WindowLocation;
+    navigate: NavigateFn;
+}
+
+export class Match<TParams> extends React.Component<MatchProps<TParams>> {}
+
+export type NavigateFn = (to: string, options?: NavigateOptions<{}>) => void;
+
+export interface NavigateOptions<TState> {
+    state?: TState;
+    replace?: boolean;
+}
+
+export class Location extends React.Component<LocationContext> {}
```

Comment:
Hi, I think this should be something like:
```ts
export class Location extends React.Component<{
  children: React.ReactNode | LocationProviderRenderFn;
}> {}
```

Using Location now throws an error on expecting props of location/navigation, but it should only expect children (https://github.com/reach/router/blob/master/src/index.js#L44)